09_simulations/csd_simulations.py

```diff
@@ -0,0 +1,174 @@
+
+    for i, beta_i in enumerate(betas):
+        ys_i = np.random.choice([-1, 1], size=(n_per_domain, 1))
+        xs_i = (
+            np.tile(ys_i, (1, p)) *
+            np.tile((np.array([beta_i]) @ z), (n_per_domain, 1))
+        ) + (np.random.normal(scale=noise_scale, size=(n_per_domain, p)))
+        if xs is None:
+            xs = xs_i
+        else:
+            xs = np.concatenate((xs, xs_i))
+        if ys is None:
+            ys = ys_i
+        else:
+            ys = np.concatenate((ys, ys_i))
+        
+    return xs, ys
+
+
+def simulate_no_csd_large_z(n_domains, n_per_domain, p, k, noise_scale=1.):
+    """Simulate data from a latent variable model with multiple domains
+    and multiple dimensions in the latent space.
+    """
+    xs, ys = None, None
+    assert k < p, 'latent dimension must be smaller than # of features'
+    # generate orthogonal matrix and take the first k vectors as latent vars
+    z = ortho_group.rvs(p)[:k, :]
+    betas = np.random.uniform(-1, 2, size=(n_domains, k))
+
+    for i in range(n_domains):
+        beta_i = betas[[i], :]
+        ys_i = np.random.choice([-1, 1], size=(n_per_domain, 1))
+        xs_i = (
+            np.tile(ys_i, (1, p)) *
+            np.tile((np.array(beta_i) @ z), (n_per_domain, 1))
+        ) + (np.random.normal(scale=noise_scale, size=(n_per_domain, p)))
+        if xs is None:
+            xs = xs_i
+        else:
+            xs = np.concatenate((xs, xs_i))
+        if ys is None:
+            ys = ys_i
+        else:
+            ys = np.concatenate((ys, ys_i))
+        
+    return xs, ys
+
+
+def simulate_csd(n_domains, n_per_domain, p, k, noise_scale=1.):
+    """Simulate data from the CSD ('common-specific decomposition') model,
+    with uncorrelated noise within domains.
+
+    CSD and its generative model are described in more detail here:
+    http://proceedings.mlr.press/v119/piratla20a/piratla20a.pdf
+    """
+    xs, ys = None, None
+    assert k < p, 'latent dimension must be smaller than # of features'
+
+    # generate orthogonal matrix
+    z = ortho_group.rvs(p)
+
+    # take the first vector as the common latent component
+    z_c = z[:1, :]
+    # take the next k vectors as the specific latent components
+    z_s = z[1:k+1, :]
+
+    betas = np.random.uniform(-1, 2, size=(n_domains, k))
+
+    for i in range(n_domains):
+        beta_i = betas[[i], :]
+        ys_i = np.random.choice([-1, 1], size=(n_per_domain, 1))
+        xs_i = (
+            np.tile(ys_i, (1, p)) *
+            np.tile(z_c + (np.array(beta_i) @ z_s), (n_per_domain, 1))
+        ) + (np.random.normal(scale=noise_scale, size=(n_per_domain, p)))
+        if xs is None:
+            xs = xs_i
+        else:
+            xs = np.concatenate((xs, xs_i))
+        if ys is None:
+            ys = ys_i
+        else:
+            ys = np.concatenate((ys, ys_i))
+        
+    return xs, ys
+
+
+def simulate_csd_corr(n_domains, n_per_domain, p, k, corr_top=1., diag=None):
```

Comment:
You could consider merging this and `simulate_csd_corr` with a parameter of whether or not to use correlated noise, since the rest of the code seems similar.